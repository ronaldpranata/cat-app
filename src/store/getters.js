export default {
  getCatBreedColumn() {
    return ['Image', 'Name'];
  },
  getCatBreed(state) {
    return state.catBreeds;
  },
  getCatDetail(state) {
    return state.catDetail;
  },
  getOriginMap(state) {
    return state.catOriginMap;
  },
};
