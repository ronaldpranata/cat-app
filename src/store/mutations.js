export default {
  SET_CAT_BREEDS: (state, data) => {
    state.catBreeds = data;
  },
  SET_MAP_LOCATION: (state, data) => {
    state.catOriginMap = data;
  },
};
