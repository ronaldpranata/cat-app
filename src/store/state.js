export default () => ({
  catDetail: null,
  catBreeds: [],
  catOriginMap: null,
});
