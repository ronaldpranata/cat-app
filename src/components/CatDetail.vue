<template>
  <div class="cat">
    <a class="bold clickable" @click="$router.go(-1)">&laquo; back</a><br />
    <h1 class="title">{{ t('catDetail') + " " + cat?.name }}</h1>
    <img :src="cat?.image?.url" class="cat-image" />
    <table>
      <tr>
        <td class="bold">{{ t('description') }}</td>
        <td class="detail description" v-html="cat?.description"></td>
      </tr>
      <tr>
        <td class="bold">{{ t('maximumWeight') }}</td>
        <td class="detail weight">
          <div v-for="(weight, key) in cat?.weight" :key="key">
            {{ key }} : {{ weight }}
          </div>
        </td>
      </tr>
      <tr>
        <td class="bold">{{ t('averageLifespan') }}</td>
        <td class="detail life-span">{{ cat?.life_span }}</td>
      </tr>

      <tr>
        <td class="bold">{{ t('origin') }}</td>
        <td class="detail gmap">
          <GMapMap :center="center" v-if="center" :zoom="3" map-type-id="terrain" style="width: 500px; height: 300px">
            <GMapCluster v-if="markers.length>0">
              <GMapMarker :key="index" v-for="(m, index) in markers" :position="m.position" :clickable="true"
                :draggable="true" @click="center = m.position" />
            </GMapCluster>
          </GMapMap>
        </td>
      </tr>
    </table>
  </div>
</template>

<script>
import { defineComponent } from 'vue';
import { useI18n } from 'vue-i18n';
import { GET_ORIGIN_LOCATION } from '@/store/action_types';

export default defineComponent({
  name: 'CatDetail',
  computed: {
    center() {
      return this.$store.getters.getOriginMap?.results[0]?.geometry?.location;
    },
    markers() {
      return [
        {
          position: this.center,
        },
      ];
    },
  },
  props: {
    cat: Object,
  },
  setup() {
    const { t } = useI18n();
    return { t };
  },
  async created() {
    await this.$store.dispatch(GET_ORIGIN_LOCATION, { address: this.cat.origin });
  },
});
</script>
<style scoped lang='scss'>
.cat {
  max-width: 700px;
  margin: auto;

  .cat-image {
    max-width: 200px;
  }

  .bold {
    font-weight: bold;
  }

  .detail {
    text-align: left;
  }
}
</style>
