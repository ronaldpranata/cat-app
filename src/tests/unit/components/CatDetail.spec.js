import { shallowMount } from '@vue/test-utils';
import { createStore } from 'vuex';
import CatDetail from '@/components/CatDetail.vue';
import { createI18n } from 'vue-i18n';
import state from '@/store/state';
import mutations from '@/store/mutations';
import getters from '@/store/getters';
import {
  GET_ORIGIN_LOCATION,
} from '@/store/action_types';

const i18n = createI18n({
  // vue-i18n options here ...
});

const map = {
  results: [
    {
      address_components: [
        {
          long_name: 'United Kingdom',
          short_name: 'GB',
          types: ['country', 'political'],
        },
      ],
      formatted_address: 'United Kingdom',
      geometry: {
        bounds: {
          northeast: {
            lat: 60.91569999999999,
            lng: 33.9165549,
          },
          southwest: {
            lat: 34.5614,
            lng: -8.8988999,
          },
        },
        location: {
          lat: 55.378051,
          lng: -3.435973,
        },
        location_type: 'APPROXIMATE',
        viewport: {
          northeast: {
            lat: 61.5471111,
            lng: 9.584415699999999,
          },
          southwest: {
            lat: 47.5554486,
            lng: -18.5319589,
          },
        },
      },
      place_id: 'ChIJqZHHQhE7WgIReiWIMkOg-MQ',
      types: ['country', 'political'],
    },
  ],
  status: 'OK',
};

const actions = {
  async [GET_ORIGIN_LOCATION]({ commit }) {
    try {
      const response = await new Promise((resolve) => {
        resolve(map);
      });
      commit('SET_MAP_LOCATION', response?.data);
      return response?.data;
    } catch (error) {
      throw new Error(error);
    }
  },
};
const store = createStore({
  state,
  actions,
  mutations,
  getters,
});
const factory = () => shallowMount(CatDetail, {
  propsData: {
    cat: {
      image: {
        url: 'imgurl',
      },
      name: 'catName',
      description: 'catdescription',
      weight: {
        test: 'test',
      },
      life_span: '3-4',
      origin: 'United States',
    },
  },
  global: {
    plugins: [i18n],
    provide: {
      store,
    },
  },

});

describe('CatDetail', () => {
  test('mounts properly', () => {
    const wrapper = factory();
    expect(wrapper.vm).toBeTruthy();
  });

  test('renders properly', () => {
    const wrapper = factory();
    expect(wrapper.html()).toMatchSnapshot();
  });
});
