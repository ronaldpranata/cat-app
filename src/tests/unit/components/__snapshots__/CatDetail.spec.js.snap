// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CatDetail renders properly 1`] = `
<div class="cat"><a class="bold clickable">Â« back</a><br>
  <h1>catDetail catName</h1><img src="imgurl" class="cat-image">
  <table>
    <tr>
      <td class="bold">description</td>
      <td class="detail">catdescription</td>
    </tr>
    <tr>
      <td class="bold">maximumWeight</td>
      <td class="detail">
        <div>test : test</div>
      </td>
    </tr>
    <tr>
      <td class="bold">averageLifespan</td>
      <td class="detail">3-4</td>
    </tr>
    <tr>
      <td class="bold">origin</td>
      <td class="detail">
        <!--v-if-->
      </td>
    </tr>
  </table>
</div>
`;
